llvm/lib/Target/AArch64/AArch64InstrInfo.cpp

```diff
@@ -10179,11 +10179,33 @@ AArch64InstrInfo::getOutliningTypeImpl(const MachineModuleInfo &MMI,
       return outliner::InstrType::Illegal;
   }
 
-  // Special cases for instructions that can always be outlined, but will fail
-  // the later tests. e.g, ADRPs, which are PC-relative use LR, but can always
-  // be outlined because they don't require a *specific* value to be in LR.
-  if (MI.getOpcode() == AArch64::ADRP)
+  // An ADRP instruction referencing a GOT should not be outlined.
+  // This is to avoid splitting ADRP/(LDR/ADD/etc.) pair into different
+  // functions which can lead to linker ICF merging sections incorrectly.
+  if (MI.getOpcode() == AArch64::ADRP) {
+    bool IsPage = (MI.getOperand(1).getTargetFlags() & AArch64II::MO_PAGE) != 0;
+    bool IsGot = (MI.getOperand(1).getTargetFlags() & AArch64II::MO_GOT) != 0;
+    if (IsPage && IsGot)
+      return outliner::InstrType::Illegal;
+
+    // Special cases for instructions that can always be outlined, but will fail
+    // the later tests. e.g, ADRPs, which are PC-relative use LR, but can always
+    // be outlined because they don't require a *specific* value to be in LR.
     return outliner::InstrType::Legal;
+  }
+
+  // Similarly, any user of ADRP instruction referencing a GOT should not be
+  // outlined. It's hard/costly to check exact users of ADRP. So we use check
+  // all operands and reject any that's a page offset and references a GOT.
+  const auto &F = MI.getMF()->getFunction();
+  for (const auto &MO : MI.operands()) {
+    bool IsPageOff = (MO.getTargetFlags() & AArch64II::MO_PAGEOFF) != 0;
+    bool IsGot = (MO.getTargetFlags() & AArch64II::MO_GOT) != 0;
+    if (IsPageOff && IsGot &&
+        (MI.getMF()->getTarget().getFunctionSections() || F.hasComdat() ||
```

Comment:
The comment above says:
```
So we use check all operands and reject any that's a page offset and references a GOT.
```` 
This part seems to further constrain to only functions that are, roughly speaking, compiled in their own section.

However the ADRP is not constrained. Can you explain why this is needed? If there is a good reason, I think we should have a test case.

Could we end up with a case where there is an attempt to outline an ADRP, ADD sequence from .text where the ADRP is illegal, but the ADD is not, and this causes problems?

Reply:
Hmm. My code was partially motivated by @lenary's code pointer to RISCV backend here: https://github.com/llvm/llvm-project/issues/131660#issuecomment-2731554890 

Maybe I am missing something there but now that I look at it again, I agree that this should just be 
```
if (isPageOff && IsGot) { ... }
```

Reply:
The additional conditions are to detect the cases where the outlined function will be in a separate section to the original function. If you're outlining from a function to create a function in the same section, then you shouldn't need to worry about ICF (which only works with -ffunction-sections, iirc). This is the approach we made on RISC-V, but I note that ADRP does work slightly differently to AUIPC.

Reply:
Thanks for the additional context. I had made the assumption that the outliner could always create a new section when outlining, but if it is in the same section then I agree we won't see the ICF problem.

I think it would be worth a quick comment to explain.

Reply:
The outliner creates new functions (there will be a little LLVM IR function stub for them, because there has to be, but the MachineFunction contents is the important bit), which might inherit section-related metadata, but the section creation still happens as usual in the AsmPrinter, which is at the end of the codegen pipeline.

Reply:
I moved this logic to a different function. Now we don't create outlining candidate if adrp pair is split. Checking for function-sections is not relevant anymore.

Marking this as resolved